panic (trackercounters flatten assertion error) when running e2e tests for pidwatcher with multiple trackers

When running e2e tests for pidwatcher with multiple trackers ( refer to: https://github.com/intel/memtierd/pull/5/), the following panic err will be raised.

For all the pidwatchers (PidWatcherProc, PidWatcherCgroups, pidlist), this issue can be reproduced. The following err is for when pidwatcher is pidlist.

DEBUG: memtier TrackerIdlePage: AddPids([10412 10427 10442])
DEBUG: memtier TrackerSoftDirty: online
DEBUG: memtier TrackerIdlePage: online
DEBUG: memtier TrackerSoftDirty: AddPids([10412 10427 10442])
memtierd> DEBUG: memtier Mover: online
DEBUG: memtier PolicyAge: online
DEBUG: memtier PolicyAge: offline
panic: trackercounters flatten assertion error: not sorted by addresses

goroutine 9 [running]:
github.com/intel/memtierd/pkg/memtier.(*TrackerCounters).Flattened(0xc000065c50, 0x0, 0x0)
	/root/go/src/github.com/intel/memtierd/pkg/memtier/tracker.go:254 +0x1067
github.com/intel/memtierd/pkg/memtier.(*TrackerMulti).GetCounters(0xc000016200?)
	/root/go/src/github.com/intel/memtierd/pkg/memtier/tracker_multi.go:108 +0x14d
github.com/intel/memtierd/pkg/memtier.(*PolicyAge).loop(0xc0000164c0)
	/root/go/src/github.com/intel/memtierd/pkg/memtier/policy_age.go:459 +0x23a
created by github.com/intel/memtierd/pkg/memtier.(*PolicyAge).Start
	/root/go/src/github.com/intel/memtierd/pkg/memtier/policy_age.go:307 +0x12a
2023/07/07 12:47:00 socat[10498] E write(1, 0x564815a6d000, 53): Broken pipe
2023/07/07 12:47:03 socat[10512] E write(1, 0x564815a6d000, 53): Broken pipe
2023/07/07 12:47:05 socat[10526] E write(1, 0x564815a6d000, 53): Broken pipe
2023/07/07 12:47:07 socat[10540] E write(1, 0x564815a6d000, 53): Broken pipe
2023/07/07 12:47:09 socat[10554] E write(1, 0x564815a6d000, 53): Broken pipe
2023/07/07 12:47:11 socat[10568] E write(1, 0x564815a6d000, 53): Broken pipe

--- a/pkg/memtier/tracker.go
+++ b/pkg/memtier/tracker.go
@@ -192,8 +192,9 @@ func flattenDefaultUnion(tc0, tc1 TrackerCounter) TrackerCounter {
 
 // Flattened returns tracker counters with overlapping parts squashed.
 // Parameters:
-//   cut - returns new tc that is a cut from the address range part of tc0.
-//   union - returns new tc that represents union of tc0 and tc1.
+//
+//	cut - returns new tc that is a cut from the address range part of tc0.
+//	union - returns new tc that represents union of tc0 and tc1.
 func (tcs *TrackerCounters) Flattened(cut func(tc0 TrackerCounter, ar *AddrRange) TrackerCounter, union func(tc0, tc1 TrackerCounter) TrackerCounter) *TrackerCounters {
 	tcs.SortByAddr()
 	// Invariants:
@@ -210,9 +211,10 @@ func (tcs *TrackerCounters) Flattened(cut func(tc0 TrackerCounter, ar *AddrRange
 	for _, tc := range *tcs {
 		tcStartAddr := tc.AR.Ranges()[0].Addr()
 		tcEndAddr := tc.AR.Ranges()[0].EndAddr()
+		tcPid := tc.AR.Pid()
 
 		if len(flatTcs) == 0 ||
-			flatTcs[len(flatTcs)-1].AR.Pid() != tc.AR.Pid() ||
+			flatTcs[len(flatTcs)-1].AR.Pid() != tcPid ||
 			flatTcs[len(flatTcs)-1].AR.Ranges()[0].EndAddr() <= tcStartAddr {
 			flatTcs = append(flatTcs, tc)
 			continue
@@ -224,7 +226,7 @@ func (tcs *TrackerCounters) Flattened(cut func(tc0 TrackerCounter, ar *AddrRange
 		oltci := len(flatTcs) - 1
 		for oltci >= 0 {
 			prevEndAddr := flatTcs[oltci].AR.Ranges()[0].EndAddr()
-			if prevEndAddr < tcStartAddr {
+			if flatTcs[oltci].AR.Pid() != tcPid || prevEndAddr < tcStartAddr {
 				// No overlap at this index, the
 				// previous index was the last one.
 				oltci++

